File: app/src/main/java/com/tetsworks/pocketlive/audio/Sequencer.java
package com.tetsworks.pocketlive.audio;

import android.os.Handler;
import android.os.Looper;
import com.tetsworks.pocketlive.model.Track;
import java.util.ArrayList;
import java.util.List;

public class Sequencer {

    public interface OnStepChangedListener {
        void onStepChanged(int step);
    }

    private final AudioEngine audioEngine;
    private List<Track> tracks = new ArrayList<>();
    private int bpm = 120;
    private int stepCount = 16;
    private int currentStep = -1;
    private boolean playing = false;
    private long stepIntervalMs = 60_000L / 120 / 2;
    private OnStepChangedListener listener;

    private final Handler handler = new Handler(Looper.getMainLooper());
    private final Runnable tickRunnable = new Runnable() {
        @Override
        public void run() {
            if (!playing) return;
            currentStep = (currentStep + 1) % stepCount;
            if (listener != null) listener.onStepChanged(currentStep);
            for (Track t : tracks) {
                if (t.getSteps()[currentStep]) audioEngine.play(t);
            }
            handler.postDelayed(this, stepIntervalMs);
        }
    };

    public Sequencer(AudioEngine audioEngine) {
        this.audioEngine = audioEngine;
    }

    public void setTracks(List<Track> tracks) { this.tracks = tracks; }

    public void setBpm(int bpm) {
        this.bpm = Math.max(40, Math.min(300, bpm));
        this.stepIntervalMs = 60_000L / this.bpm / 2;
    }

    public int getBpm() { return bpm; }
    public int getCurrentStep() { return currentStep; }
    public boolean isPlaying() { return playing; }
    public void setOnStepChangedListener(OnStepChangedListener l) { this.listener = l; }

    public void play() {
        if (playing) return;
        playing = true;
        handler.post(tickRunnable);
    }

    public void pause() {
        playing = false;
        handler.removeCallbacks(tickRunnable);
    }

    public void stop() {
        playing = false;
        handler.removeCallbacks(tickRunnable);
        currentStep = -1;
        if (listener != null) listener.onStepChanged(-1);
    }
}

File: app/src/main/java/com/tetsworks/pocketlive/audio/Sequencer.kt
package com.tetsworks.pocketlive.audio

import android.os.Handler
import android.os.Looper
import com.tetsworks.pocketlive.model.Track

class Sequencer(private val audioEngine: AudioEngine) {

    var bpm: Int = 120
        set(value) {
            field = value.coerceIn(40, 300)
            stepIntervalMs = (60_000L / field / 2)
        }

    var stepCount: Int = 16
    var currentStep: Int = -1
        private set

    var isPlaying: Boolean = false
        private set

    var onStepChanged: ((step: Int) -> Unit)? = null

    private var stepIntervalMs: Long = (60_000L / 120 / 2)
    private val handler = Handler(Looper.getMainLooper())
    private var tracks: List<Track> = emptyList()

    private val tickRunnable = object : Runnable {
        override fun run() {
            if (!isPlaying) return
            currentStep = (currentStep + 1) % stepCount
            onStepChanged?.invoke(currentStep)
            tracks.forEach { track ->
                if (track.steps[currentStep]) {
                    audioEngine.play(track)
                }
            }
            handler.postDelayed(this, stepIntervalMs)
        }
    }

    fun setTracks(tracks: List<Track>) {
        this.tracks = tracks
    }

    fun play() {
        if (isPlaying) return
        isPlaying = true
        handler.post(tickRunnable)
    }

    fun pause() {
        isPlaying = false
        handler.removeCallbacks(tickRunnable)
    }

    fun stop() {
        isPlaying = false
        handler.removeCallbacks(tickRunnable)
        currentStep = -1
        onStepChanged?.invoke(-1)
    }
}
